File: geekbang/algorithm-models/3.5-子集递归模板/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        ans = new ArrayList<List<Integer>>();
        s = new ArrayList<Integer>();
        findSubsets(nums, 0);
        return ans;
    }


    private void findSubsets(int[] nums, int index) {
        if (index == nums.length) {
            ans.add(new ArrayList<Integer>(s));
            return;
        }
        findSubsets(nums, index + 1);
        s.add(nums[index]);
        findSubsets(nums, index + 1);
        s.remove(s.size() - 1);
    }


    private List<List<Integer>> ans;
    private List<Integer> s;
}

File: geekbang/algorithm-models/3.5-子集递归模板/subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        findSubsets(nums, 0);
        return ans;
    }


    void findSubsets(vector<int>& nums, int index) {
        if (index == nums.size()) {
            ans.push_back(s);
            return;
        }
        findSubsets(nums, index + 1);
        s.push_back(nums[index]);
        findSubsets(nums, index + 1);
        s.pop_back();
    }


private:
    vector<vector<int>> ans;
    vector<int> s;
};
